static/js/src/advantage/credentials/dashboard/components/BadgeTracking/BadgeTracking.tsx

```diff
@@ -0,0 +1,220 @@
+import { useState, useMemo, useEffect } from "react";
+import { Pagination, Spinner, Notification } from "@canonical/react-components";
+import { useQuery, useQueryClient } from "@tanstack/react-query";
+import { getIssuedBadgesCredly } from "../../api/queryFns";
+import { CredlyMetadata, CredlyBadge } from "../../utils/types";
+import { ModularTable } from "@canonical/react-components";
+import { upperCaseFirstChar } from "../../utils/common";
+
+type APIResponse = {
+  data: CredlyBadge[];
+  metadata: CredlyMetadata;
+  message?: string;
+};
+
+const BadgeTracking = () => {
+  const queryClient = useQueryClient();
+  const [page, setPage] = useState<number>(0);
+  const [fetchedPages] = useState(new Set([1]));
+  const [cachedData, setCachedData] = useState<Record<number, APIResponse>>({});
+  const filter = "state::pending,rejected,revoked";
+  // const [sort, setSort] = useState(null);
+  const sort = null;
+
+  const onSuccess = (newData: APIResponse) => {
+    if (newData && newData.data) {
+      setCachedData((prev) => ({
+        ...prev,
+        [page]: newData,
+      }));
+    }
+  };
+
+  const { isLoading, isError, data, isFetching } = useQuery<APIResponse>({
+    queryKey: ["issuedBadges", filter, sort, page],
+    queryFn: () => getIssuedBadgesCredly(filter, sort, page, onSuccess),
+  });
```

Comment:
Something like this? 

```suggestion
  const { isLoading, isError, data, isFetching } = useQuery<APIResponse>({
    queryKey: ["issuedBadges", filter, sort, page],
    queryFn: () => getIssuedBadgesCredly(filter, sort, page),
    onSuccess: (newData)=> {
      if (newData && newData.data) {
        setCachedData((prev) => ({
          ...prev,
          [page]: newData,
        }));
      }
    },
  });
```

Reply:
Oh right, but I think its not the best practice side effects are made in the query function according to this docs - [on success](https://tanstack.com/query/latest/docs/framework/react/guides/migrating-to-react-query-4#onsuccess-is-no-longer-called-from-setquerydata) and [query functions](https://tanstack.com/query/latest/docs/framework/react/guides/query-functions). 

I think side effects and state updates should be handled outside of query function. 

Can we use useEffect instead?
```
useEffect(() => {
    if (data && data.data) {
      setCachedData((prev) => ({
        ...prev,
        [page]: data,
      }));
    }
  }, [data, page]);
```
I might be wrong though. What do you think?